### DISCUSSION TITLE
Expose `pyodide.loadPackage` arguments in Pyscript, to provide better loading UX

### DISCUSSION THREAD

**Discussion by brunokim** [pyscript/pyscript#2160] [Proposals]
I've seen console logs like 'Loading micropip', and traced them to pyodide's `loadPackage`. This function may take callbacks for reporting messages and errors, which by default calls `console.log`:

https://github.com/pyodide/pyodide/blob/c48d3c942046ed1c3f39676e9db5a8bbca319b92/src/js/load-package.ts#L157-L160

I'd like to expose these logs to the user, to show some progress other than just a single "Loading...". For that, all its usage within pyscript would need to expose the callbacks to its callers, e.g., in instances of

https://github.com/search?q=org%3Apyscript+loadPackage&type=code

**Comment by WebReflection** [comment]
the question is *how* ... and *where* would you provide your own listeners ... unless you are asking us to emit at the window level some event per each logged action which feels like the right approach since we already dispatch a `py:done` event + others ... would that work for you? I kinda like the idea, but I don't see many ways to enable it beside events (and I think we should keep showing logs regardless for everyone else not listening to such events).

**Reply by WebReflection** [reply]
current PyScript events https://docs.pyscript.net/2024.8.2/faq/#pyscript-events

**Reply by brunokim** [reply]
That's a good solution! I don't mind if the logging keeps happening on the console.

**Comment by WebReflection** [comment]
FYI this is being tackled in here https://github.com/pyscript/pyscript/pull/2162

**Reply by WebReflection** [reply]
to provide more context `loadPackage` callbacks don't get propagated to *micropip* so that's a dead-end to know what's going on when you have packages around, as those callbacks would never tell you anything micropip is doing or signaling while bootstrapping (cc @hoodmane) ... I ended up mokey-patching temporarily the `console.log` while bootstrapping to be able to actually fetch all those logs nobody asked for or can easily hook into and the result is actually looking pretty good.

Events are triggered on the main thread even from workers so there is no disambiguation needed: you listen to `py:progress` event on the page, you get what you asked for ... as example, this is having *matplotlib* as package and all events appended to the body before a `print(sys.version)` (from a worker, which dispatches on the main ... it works on main too):

![Screenshot from 2024-09-12 17-22-08](https://github.com/user-attachments/assets/b4a19271-d712-40b1-93ff-a6d92ba028d8)

**Reply by brunokim** [reply]
How did you patch `console.log`? I've attempted to set `window.console = myConsole` in a <script> loading before all of them, but this wasn't enough to intercept the calls from WASM.

**Reply by WebReflection** [reply]
it's always a hack to patch globals so I am not super happy to share the how because if done in a bad way everything else would break but the nitty gritty is that you don't reassign window console, you overload `console.log` and friends but be sure to put back the original version or bad things could happen.

**Comment by WebReflection** [comment]
a showcase of progress events live here: https://agiammarchi.pyscriptapps.com/kmeans-in-panel-copy/latest/

**Reply by WebReflection** [reply]
I think we can close this as it's landing within the next release (soon)
